### DISCUSSION TITLE
Latency Testing per Dev Journal

### DISCUSSION THREAD

**Discussion by GamingNJncos** [darthcloud/BlueRetro#330] [General]
Could use some general pointers on latency collection as noted here as it relates to stand alone devices: https://hackaday.io/project/170365-blueretro/log/187443-2020-12-26-update-latency-tests-release-v010

Testing with a stand alone ESP wroom devkit, after flashing the BlueRetro_pad_test.bin I noted that GPIO26 appears to be held high at all times.  Digging around, I believe this might be related to use on Dreamcast as it's used for Player4.  

I've confirmed that the Mapping Config setting is Output1 All Latency Test but can't seem to get IO26 to print on button over BT.  I can see the button being detected as pressed on the TX and RX for BlueRetro shows the packets coming in, just not able to toggle GPIO26 as expected. Edit:  Should note that I do see the pin held high with no controllers connected.

Is there a specific firmware or some other pins I might need to pull high or low?

**Comment by darthcloud** [comment]
~Its a special firmware that isn't build by default.~

~Will try provide it tomorrow.~

Edit: This is false see below

**Reply by darthcloud** [reply]
Correction, you need to use the parallel_1P_external.bin

And use a presets or mapping that makes all buttons toggle the same pin. It used to be IO26 but the recent buttons change might have change that.

Will look tomorrow.

**Reply by darthcloud** [reply]
BlueRetro need to be run standalone for that test (no console) only the arduino is plug to io26.

**Reply by darthcloud** [reply]
The presets is called "Latency Test" and I confirmed the IO26 is the right pin.

So using this arduino sketch: https://pastebin.com/ktqn7izx you plug arduino pin 2 to ESP32 IO26 and connect arduino pin 5 to one of the buttons of your controller.

And you let it run 10K+ iteration.

**Reply by GamingNJncos** [reply]
Thanks for clarifying, I got it working as advised. 

Useful note - I am using a pro micro for the Arduino and had to move from pin 5 to pin 10 for the button trigger as the close proximity for the two pins would trigger the interrupt "magically" due to noise/interference. As a result it printed *really* low latency, but it was bad data.

**Reply by GamingNJncos** [reply]
One Item I wanted to call out I noted that there are 2 ways a user could try this but only one of the 2 works. I tested on 1.7.1 but noted it on earlier version also.

Providing Screenshots for clarity
![Screenshot_20220726-131434_Chrome](https://user-images.githubusercontent.com/106001964/181081926-28c2eff5-fb45-421e-913e-e1e7c2fb68aa.jpg)
![Screenshot_20220726-131543_Chrome](https://user-images.githubusercontent.com/106001964/181081939-41ba75ab-03e3-4805-a29f-11b314f51ceb.jpg)
